dropshot_endpoint/src/lib.rs

```diff
@@ -219,6 +220,20 @@ fn do_channel(
                 ));
             }
 
+            // Historically, we required that the `WebsocketConnection` argument
+            // be first after the `RequestContext`.  However, we also require
+            // that any exclusive extractor (which includes the
+            // `WebsocketUpgrade` argument that we put in its place) appears
+            // last.  We replaced the type above, but now we need to put it in
+            // the right spot.
+            sig.inputs = {
+                let mut input_pairs =
+                    sig.inputs.clone().into_pairs().collect::<Vec<_>>();
+                let second_pair = input_pairs.remove(1);
+                input_pairs.push(second_pair);
+                input_pairs.into_iter().collect()
+            };
```

Comment:
I'm curious what people think about this.  The comment basically summarizes it: we previously said that a `WebsocketConnection` has to be the first extractor.  As part of this PR, I decided that exclusive extractors (which would include this one) have to be _last_.  I did that because I think it's natural for the extractors to appear in the order they look at parts of the HTTP request, and exclusive extractors generally look at the request body (the last thing).

There's not actually a conflict here because this macro already translates the user's argument list into what the `endpoint` macro expects, so we can shuffle the arguments around like this.  But it's kind of annoying that we have to do this, and a user might notice the weirdness that their `WebsocketConnection` has to appear first, but other exclusive extractors have to appear last.

We could:

- do nothing (what we're doing here).  Downside: this ugliness, and maybe slightly confusing for people?
- make `WebsocketConnection` appear last.  Downside: breaking change, though trivial to fix.
- make exclusive extractors appear first instead.  Downside: less intuitive for the much more common case

Thoughts @ahl @lifning ?

Reply:
I like making `WebsocketConnection` appear last -- it's a breaking change but I think it's justified by the overall direction of the PR.